### PR TITLE
UDFs for comparator and math expression

### PR DESCRIPTION
**Pull request by navis**
StringComaprator and Function in math expression should be extensible via druid module system. I've did best effort with my limited jackson knowledge but I believe there would be much better way than this. In that case, I'm welcome to anyone to take this issue.

### CODE REVIEW

**Comment by gianm** [comment]
note, #3270 will result in extensibility for StringComparator.

**Comment by navis** [comment]
@gianm we can register it to jackson but cannot use it because new comparator is not in @JsonCreator of StringComparator.

**Comment by gianm** [comment]
@navis I think the mechanism from #3270 should allow a custom StringComparator to be specified like `"ordering" : {"type":"myComparator"}` if it's registered as a subtype. `"ordering" : "myComparator"` would not work, that'd only be for the "built-in" comparators, but I think that's fine?

Behavior here is similar to QueryGranularity, where `"hour"` and `{"type" : "duration", "duration" : 3600000 }` are equivalent.

**Comment by navis** [comment]
@gianm Ah, sorry. I've missed OrderByColumnSpec is changed to accept StringComparator directly. Then we can use of it, thanks.
After that is merged I can modify this just for expression functions.

**Comment by navis** [comment]
@gianm got some complaints on `"ordering" : {"type":"myComparator"}`. so we've decided to use this :)

**Comment by gianm** [comment]
@navis Could you please remove the comparator stuff, or at least split it into a different PR? (doesn't seem related to the math expression stuff)

**Comment by navis** [comment]
@gianm removed that. thanks.

File: examples/src/main/java/io/druid/examples/function/GuavaDoubleMath.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+import com.google.common.math.DoubleMath;
+import io.druid.math.expr.BuiltinFunctions.SingleParam;
+import io.druid.math.expr.BuiltinFunctions.TripleParam;
+import io.druid.math.expr.ExprEval;
+import io.druid.math.expr.Function;
+
+/**
+ */
+public class GuavaDoubleMath implements Function.Library
```
Comment (fjy) [inline examples/src/main/java/io/druid/examples/function/GuavaDoubleMath.java]: DruidDoubleMath? :P

File: common/src/main/java/io/druid/math/expr/BuiltinFunctions.java [hunk 2]
```diff
@@ -0,0 +1,809 @@
...
+import org.joda.time.DateTime;
+import org.joda.time.format.DateTimeFormat;
+import org.joda.time.format.DateTimeFormatter;
+import org.joda.time.format.ISODateTimeFormat;
+
+import java.util.List;
+
+/**
+ */
+public interface BuiltinFunctions extends Function.Library
```
Comment (fjy) [inline common/src/main/java/io/druid/math/expr/BuiltinFunctions.java]: how about DruidMathFunction as a name?

File: common/src/main/java/io/druid/math/expr/Function.java [hunk 3]
```diff
@@ -40,761 +36,7 @@
...
-    public ExprEval apply(List<Expr> args, Expr.ObjectBinding bindings)
-    {
-      if (args.size() != 2) {
-        throw new IAE("function 'nvl' needs 2 arguments");
-      }
-      final ExprEval eval = args.get(0).eval(bindings);
-      return eval.isNull() ? args.get(1).eval(bindings) : eval;
-    }
+  // implement this and register to jackson module to extend expression function (see ExamplesDruidModule)
+  interface Library {
```
Comment (fjy) [inline common/src/main/java/io/druid/math/expr/Function.java]: are there no common methods at all for expression functions?

File: examples/src/main/java/io/druid/examples/ExamplesDruidModule.java [hunk 4]
```diff
@@ -38,9 +39,8 @@
   {
     return Arrays.<Module>asList(
         new SimpleModule("ExamplesModule")
-            .registerSubtypes(
-                new NamedType(TwitterSpritzerFirehoseFactory.class, "twitzer")
-            )
+            .registerSubtypes(new NamedType(TwitterSpritzerFirehoseFactory.class, "twitzer"))
+            .registerSubtypes(new NamedType(GuavaDoubleMath.class, "guavaDoubleMath"))
```
Comment (fjy) [inline examples/src/main/java/io/druid/examples/ExamplesDruidModule.java]: i'd almost want this doubleMath extension just to be a separate extension in extensions-contrib so folks can learn to write their own math extensions rather than have the code druid-examples

File: examples/src/main/java/io/druid/examples/function/GuavaDoubleMath.java (hunk 1 above)
Comment (jon-wei) [inline examples/src/main/java/io/druid/examples/function/GuavaDoubleMath.java]: hm, I'm okay with calling this "Guava", to make it clear where the function implementations come from

File: processing/src/main/java/io/druid/jackson/FunctionInitializer.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+public class FunctionInitializer implements Module
+{
+  private static final Logger log = new Logger(FunctionInitializer.class);
+
+  @Inject
+  public static void init(ObjectMapper mapper)
+  {
+    log.info("finding expression functions..");
+    for (NamedType subType : resolveSubtypes(mapper, Function.Library.class)) {
+      Parser.register(subType.getType());
```
Comment (jon-wei) [inline processing/src/main/java/io/druid/jackson/FunctionInitializer.java]: More of a question than a request for change, does this result in double-registering of the BuiltInFunctions? BuiltInFunctions also extends Function.Library, but Parser has a static block that also does the registration:

```
 static {
  register(BuiltinFunctions.class);
}
```

File: processing/src/main/java/io/druid/jackson/FunctionInitializer.java [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+/**
+ */
+public class FunctionInitializer implements Module
+{
+  private static final Logger log = new Logger(FunctionInitializer.class);
+
+  @Inject
+  public static void init(ObjectMapper mapper)
+  {
+    log.info("finding expression functions..");
```
Comment (navis) [inline processing/src/main/java/io/druid/jackson/FunctionInitializer.java]: function name can be retrieved only after instantiating instance, so it not possible to separate "before" and "after" action cleanly. I think we can introduce annotation for function to describe it in class level. Would it be better?

File: examples/src/main/java/io/druid/examples/function/GuavaDoubleMath.java (hunk 1 above)
Comment (navis) [inline examples/src/main/java/io/druid/examples/function/GuavaDoubleMath.java]: Yep. I agree with @jon-wei.

File: common/src/main/java/io/druid/math/expr/Function.java (hunk 3 above)
Comment (navis) [inline common/src/main/java/io/druid/math/expr/Function.java]: It's just a marker/wrapper class to find with jackson. wish there is a better way.

File: examples/src/main/java/io/druid/examples/ExamplesDruidModule.java (hunk 4 above)
Comment (navis) [inline examples/src/main/java/io/druid/examples/ExamplesDruidModule.java]: Would it be natural to be in example module for examples? GuavaDoubleMath is good but I don't think anyone needs that as an extension.

File: processing/src/main/java/io/druid/jackson/FunctionInitializer.java (hunk 5 above)
Comment (navis) [inline processing/src/main/java/io/druid/jackson/FunctionInitializer.java]: good catch. I've excluded BuiltinFunctions in here.

File: examples/src/main/java/io/druid/examples/ExamplesDruidModule.java (hunk 4 above)
Comment (jaehc) [inline examples/src/main/java/io/druid/examples/ExamplesDruidModule.java]: If you let me barge in, I agree with @fjy view which would make folks contribte expression functions easily when they come up with a new thing. However, it would be better to put that into a separate PR.

File: common/src/main/java/io/druid/math/expr/Parser.java [hunk 7]
```diff
@@ -44,30 +42,40 @@
...
   static {
-    Map<String, Supplier<Function>> functionMap = Maps.newHashMap();
-    for (Class clazz : Function.class.getClasses()) {
+    register(BuiltinFunctions.class);
+  }
+
+  public static void register(Class parent)
+  {
+    for (Class clazz : parent.getClasses()) {
       if (!Modifier.isAbstract(clazz.getModifiers()) && Function.class.isAssignableFrom(clazz)) {
```
Comment (navis) [inline common/src/main/java/io/druid/math/expr/Parser.java]: User can make any helper classes in it. we need not to log them all.

**Comment by fjy** [comment]
👍

File: common/src/main/java/io/druid/math/expr/BuiltinFunctions.java [hunk 8]
```diff
@@ -0,0 +1,809 @@
...
+    }
+
+    protected ExprEval eval(long param)
+    {
+      return eval((double) param);
+    }
+
+    protected ExprEval eval(double param)
+    {
+      return eval((long) param);
```
Comment (gianm) [inline common/src/main/java/io/druid/math/expr/BuiltinFunctions.java]: It seems like most of the functions that only override one "eval" will choose to override the double version. How do you feel about making the double versions abstract? (to promote better compile time checking)

Comment (gianm) [inline common/src/main/java/io/druid/math/expr/BuiltinFunctions.java]: Ah, I see this is just copy and pasted code. Nevermind. Still I think would be a nice change though.

File: server/src/main/java/io/druid/initialization/Initialization.java [hunk 9]
```diff
@@ -364,7 +364,12 @@ public static Injector makeInjectorWithModules(final Injector baseInjector, Iter
       extensionModules.addModule(module);
     }
 
-    return Guice.createInjector(Modules.override(intermediateModules).with(extensionModules.getModules()));
+    return Guice.createInjector(
+        Modules.combine(
+            Modules.override(intermediateModules).with(extensionModules.getModules()),
+            new FunctionInitializer()
```
Comment (gianm) [inline server/src/main/java/io/druid/initialization/Initialization.java]: Does this need to be down here instead of in the base modules list? It feels weird to have one module that is "more special" than all the others.

File: common/src/main/java/io/druid/math/expr/Parser.java [hunk 10]
```diff
@@ -44,30 +42,40 @@
 public class Parser
 {
   private static final Logger log = new Logger(Parser.class);
-  private static final Map<String, Supplier<Function>> func;
+
+  private static final Map<String, Supplier<Function>> functions = Maps.newHashMap();
```
Comment (gianm) [inline common/src/main/java/io/druid/math/expr/Parser.java]: Does this map need to be thread safe? I think yes, since it's written to by the initialization thread and read from query threads.

**Comment by gianm** [comment]
Bumping to 0.10.1 as 0.10.0 release is imminent. @navis, please let us know if you're still interested in pursuing this patch.
